app/models/patient.rb

```diff
@@ -270,6 +273,15 @@ def confirm_appointment_after_initial_call
     end
   end
 
+  def create_random_pin
+    return if random_pin
+    25.times do
```

Comment:
I think this is still a blocker for me as written. I've softened on the pseudo-RNG here, but I'm worried about these two things:

* `raise` as a part of control flow - this is code that gets executed for every new patient, so we have to be totally sure that if there's an error it's because the CM made a mistake, not because the rails code bit it for an independent reason.
* requiring the object be in a `valid` state to escape from this function. I think that might have some weird consequences, esp early in the callbacks, or creating some kinda weird infinite loop under some circumstances.

I think the common theme here is: is there a way to just Get a number that we know is unique without having to take and cycle? Generally people solve this problem with hashing algorithms, but sha and md5, but those return strings; FarmHash returns numbers but they're huge. People also use autoincrement to solve this problem, but you've expressed concern over that being guessable.

Absent a hashing algorithm we're unlikely to find (since hashing algorithms care about a scale way bigger than we do), I wonder if something like this might work? https://docs.mongodb.com/v3.0/tutorial/create-an-auto-incrementing-field/#optimistic-loop but this is really heavy and mongo specific.

a side question: If we were going to do this in postgres, how would we do it?

Reply:
Idea 1: (recommended)
1) Choose primes p, and q as secrets, where q has 6 digits 
2) Store p and q as ENV variables and compute the caller_pin (no longer random) id as n^p mod q

Advantages: 
1) this is a bijection of the integers 1...(q-1) onto itself, so it beats the problem of sequentiality.
2) This would mean we don't need to store another DB field at all, and therefore also wouldn't have to worry about existing records.

Disadvantages:
1) Math

-----
Idea 2 (meh)
One potential solution might be something like this:
1) pre-generate a list of random numbers and save them in an array in the DB
2) whenever we need a new random number, we use the first element in that array and delete it.

The good news: can just do something like `100_000..999_999.shuffle!` to generate them, and they're guaranteed unique since they're just a permutation.

The bad news: 
1) it's a lot more work
2) if a bug, server error, whatever, caused us to reuse a number by accident, it would have drastic business consequences, so we'd probably want to do a uniqueness validation anyways, in which case we'd need a way to "try again" if this was violated, and then we're at the same problem we started with.

Reply:
Resolving conversation so we can look at this with fresh eyes